unity/Assets/Scripts/BaseFPSAgentController.cs

```diff
@@ -3516,6 +3572,265 @@ bool includeInvisible
             );
         }
 
+        protected IEnumerable<Vector3> pointsOnSurfaceOfBoxCollider(BoxCollider bc, int divisions) {
+            if (divisions < 2) {
+                throw new ArgumentException($"divisions must be >= 2 (currently equals {divisions}).");
+            }
+            Vector3 halfSize = 0.5f * bc.size;
+
+            List<float> xMinMax = new List<float> {-halfSize.x, halfSize.x};
+            List<float> yMinMax = new List<float> {-halfSize.y, halfSize.y};
+            List<float> zMinMax = new List<float> {-halfSize.z, halfSize.z};
+
+            List<float> xCenterVals = new List<float>();
+            List<float> yCenterVals = new List<float>();
+            List<float> zCenterVals = new List<float>();
+            for (int i = 1; i < divisions - 1; i++) {
+                float alpha = (2f * i / (divisions - 1f));
+                xCenterVals.Add(-halfSize.x + halfSize.x * alpha);
+                yCenterVals.Add(-halfSize.y + halfSize.y * alpha);
+                zCenterVals.Add(-halfSize.z + halfSize.z * alpha);
+            }
+
+            for (int whichX = 0; whichX < 2; whichX++) {
+                List<float> xVals = whichX == 1 ? xMinMax : xCenterVals;
+                
+                for (int whichY = 0; whichY < 2; whichY++) {
+                    List<float> yVals = whichY == 1 ? yMinMax : yCenterVals;
+
+                    for (int whichZ = 0; whichZ < 2; whichZ++) {
+                        List<float> zVals = whichZ == 1 ? zMinMax : zCenterVals;
+
+                        if (whichX + whichY + whichZ == 0) {
+                            continue;
+                        }
+
+                        # if UNITY_EDITOR
+                        Vector3? lastPoint = null;
+                        # endif
+                        foreach (float x in xVals) {
```

Comment:
For future reference you can  do `new List<float> {xVals, yVals, zVals}.CartesianProduct().Select(p => bc.center + new Vector3(p[0], p[1], p[2])` defined in our `Extensions.cs`. But I think we should keep it as it is because it's probably the fastest.

Reply:
Very cool 👍